jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCookieStore.java

```diff
@@ -0,0 +1,388 @@
+
+import org.eclipse.jetty.util.NanoTime;
+import org.eclipse.jetty.util.thread.AutoLock;
+
+/**
+ * <p>A container for {@link HttpCookie}s.</p>
+ * <p>HTTP cookies are stored along with a {@link URI} via {@link #add(URI, HttpCookie)}
+ * and retrieved via {@link #match(URI)}, which implements the path matching algorithm
+ * defined by <a href="https://www.rfc-editor.org/rfc/rfc6265">RFC 6265</a>.</p>
+ */
+public interface HttpCookieStore
+{
+    /**
+     * <p>Adds a cookie to this store, if possible.</p>
+     * <p>The cookie may not be added for various reasons; for example,
+     * it may be already expired, or its domain attribute does not
+     * match that of the URI, etc.</p>
+     * <p>The cookie is associated with the given {@code URI}, so that
+     * a call to {@link #match(URI)} returns the cookie only if the
+     * URIs match.</p>
+     *
+     * @param uri the {@code URI} associated with the cookie
+     * @param cookie the cookie to add
+     * @return whether the cookie has been added
+     */
+    public boolean add(URI uri, HttpCookie cookie);
+
+    /**
+     * @return all the cookies
+     */
+    public List<HttpCookie> all();
+
+    /**
+     * <p>Returns the cookies that match the given {@code URI}.</p>
+     *
+     * @param uri the {@code URI} to match against
+     * @return a list of cookies that match the given {@code URI}
+     */
+    public List<HttpCookie> match(URI uri);
+
+    /**
+     * <p>Removes the cookie associated with the given {@code URI}.</p>
+     *
+     * @param uri the {@code URI} associated with the cookie to remove
+     * @param cookie the cookie to remove
+     * @return whether the cookie has been removed
+     */
+    public boolean remove(URI uri, HttpCookie cookie);
+
+    /**
+     * <p>Removes all the cookies from this store.</p>
+     *
+     * @return whether the store modified by this call
+     */
+    public boolean clear();
+
+    /**
+     * <p>An implementation of {@link HttpCookieStore} that does not store any cookie.</p>
+     */
+    public static class Empty implements HttpCookieStore
+    {
+        @Override
+        public boolean add(URI uri, HttpCookie cookie)
+        {
+            return false;
+        }
+
+        @Override
+        public List<HttpCookie> all()
+        {
+            return List.of();
+        }
+
+        @Override
+        public List<HttpCookie> match(URI uri)
+        {
+            return List.of();
+        }
+
+        @Override
+        public boolean remove(URI uri, HttpCookie cookie)
+        {
+            return false;
+        }
+
+        @Override
+        public boolean clear()
+        {
+            return false;
+        }
+    }
+
+    /**
+     * <p>A default implementation of {@link HttpCookieStore}.</p>
+     */
+    public static class Default implements HttpCookieStore
+    {
+        private final AutoLock lock = new AutoLock();
+        private final Map<Key, List<HttpCookie>> cookies = new HashMap<>();
```

Comment:
Would this be better as a 
```suggestion
        private final Map<Key, PathMapper<HttpCookie>> cookies = new HashMap<>();
```

I guess that is an optimisation that can come later.

Reply:
There is no `PathMapper` class? Besides, the match is done on domains and paths, so not sure we have an existing solution for that?